src/java.base/share/classes/java/util/stream/Gatherer.java

```diff
@@ -0,0 +1,577 @@
+     */
+    @FunctionalInterface
+    @PreviewFeature(feature = PreviewFeature.Feature.STREAM_GATHERERS)
+    interface Integrator<A, T, R> {
+        /**
+         * Performs an action given: the current state, the next element, and
+         * a downstream object; potentially inspecting and/or updating
+         * the state, optionally sending any number of elements downstream
+         * -- and then returns whether more elements are to be consumed or not.
+         *
+         * @param state The state to integrate into
+         * @param element The element to integrate
+         * @param downstream The downstream object of this integration,
+         *                   returns false if doesn't want any more elements
+         * @return {@code true} if subsequent integration is desired,
+         *         {@code false} if not
+         */
+        boolean integrate(A state, T element, Downstream<? super R> downstream);
+
+        /**
+         * Factory method for turning Integrator-shaped lambdas into
+         * Integrators.
+         *
+         * @param integrator a lambda to create as Integrator
+         * @return the given lambda as an Integrator
+         * @param <A> the type of elements this integrator receives
+         * @param <T> the type of initializer this integrator consumes
+         * @param <R> the type of results this integrator can produce
+         */
+        @ForceInline
+        static <A, T, R> Integrator<A, T, R> of(Integrator<A, T, R> integrator) {
+            return integrator;
+        }
+
+        /**
+         * Factory method for turning Integrator-shaped lambdas into
+         * {@link Greedy} Integrators.
+         *
+         * @param greedy a lambda to create as Integrator.Greedy
+         * @return the given lambda as a Greedy Integrator
+         * @param <A> the type of elements this integrator receives
+         * @param <T> the type of initializer this integrator consumes
+         * @param <R> the type of results this integrator can produce
+         */
+        @ForceInline
+        static <A, T, R> Greedy<A, T, R> ofGreedy(Greedy<A, T, R> greedy) {
+            return greedy;
+        }
+
+        /**
+         * Greedy Integrators consume all their input, and may only relay that
+         * the downstream does not want more elements.
+         *
+         * <p>This is used to clarify that no short-circuiting will be
+         * initiated by this Integrator, and that information can then be
+         * used to optimize evaluation.
+         *
+         * @param <A> the type of elements this greedy integrator receives
+         * @param <T> the type of initializer this greedy integrator consumes
+         * @param <R> the type of results this greedy integrator can produce
+         * @since 22
+         */
+        @FunctionalInterface
+        @PreviewFeature(feature = PreviewFeature.Feature.STREAM_GATHERERS)
+        interface Greedy<A, T, R> extends Integrator<A, T, R> { }
+    }
+}
```

Comment:
Looks like there is no newline at the end of this file. Though newline at end of file is not mandatory it avoids getting warning in some tools (such as IDE etc...)
```suggestion
}

```